### ISSUE TITLE
Call for Terminology: Assert Order and Heuristics

### ISSUE THREAD

**Issue by Heizmann** [ultimate-pa/ultimate#250]
The term **assert order** that we use internally is very confusing.
It is no so much important that we chang the assert order the order but that we do additional `check-sat`.

Conceptually, we are abstracting the trace. Unfortunately we cannot use the term *trace abstraction*.
Additionally it is noteworthy that we construct the abstraction incrementally. We always have a guarantee that the result infeasible (it might not be a strict abstraction).
Idea: *infeasibility preserving abstraction*
Do you have other ideas?

We sometimes call our algorithms for the computation of such a strategy **heuristics**. This term is too general and people might mix that up with the heuristics from the Beyer-Löwe-Wendler refinement selection.
Idea: *abstraction scheme* or *infeasibility preserving abstraction scheme*
Do you have other ideas?

**Comment by schillic** [comment]
For 1): *(incremental) trace modulation/shuffling*?
Maybe also combined with your *infeasibility preserving*...

For 2): *scheme* is good; an alternative might be *pattern*.

**Comment by Heizmann** [comment]
But there is no modulation in 1)

**Comment by alexandernutz** [comment]
Some thoughts:

 - _abstraction scheme_ seems too general.
 - On a slightly higher level, the technique is about choosing between different root causes for the infeasiblity of the trace, correct? (or more precisely: forcing the SMT-Solver to give a proof using this or that cause of infeasibility) So perhaps something like _root cause selection scheme_, _cause of infeasibility selection scheme_, ...?

**Comment by danieldietsch** [comment]
Selecting from reasons of infeasibility is also the idea behind Dirk's refinement selection. I agree that for refinement, we have more options and are therefore not directly determined by the selection of the infeasibility reason. 

In some sense what is happening here is a kind of **proof selection** (which would match the title). Because we look at multiple proofs, it could also be a **proof sampling**.

**Comment by Heizmann** [comment]
In the paper I was able to avoid the introduction of new terminology (unclear if good or not) however I think the topic is still interesting.

However, I think the discussion is going in the wrong direction. In this issue I am only asking how we should call  (1) the *abstraction of a trace* and (2) the *thing* of which e.g., OUTSIDE_LOOP_FIRST1,  INSIDE_LOOP_FIRST1, TERMS_WITH_SMALL_CONSTANTS_FIRST are instance of.

I have the impression that your answers are addressing a different question.

**Comment by danieldietsch** [comment]
I think _assertion order_ describes pretty well what is happening and I would keep it (perhaps drop _modulation_). Looking at it as an abstraction does not seem right because it is a rather specific one: an underapproximation of infeasibility. Descriptions like _trace partitioning_ (already used) or _trace reduction_ seem appropriate, but I feel like someone could imagine something different there. 

Regarding _heuristics_: What about _strategy_, _tactic_, or just saying that each different instance is just one _assertion order_  (then calling the whole thing _assertion ordering_).

**Comment by Heizmann** [comment]
I am really puzzled.

> I think assertion order describes pretty well
This is a sentence that I can understand. But I do not agree

> perhaps drop modulation
Since there is no modulation, I have no idea how I can drop it

> Looking at it as an abstraction does not seem right because it is a rather specific one
This sentence is already self-contradicting

> Descriptions like trace partitioning or trace reduction seem appropriate
> then calling the whole thing assertion ordering
What is "the whole thing"

I really have the impression that we are talking about different things.

**Comment by danieldietsch** [comment]
I am sorry that you are puzzled. 

> Since there is no modulation, I have no idea how I can drop it

In our implementation we call it assertion order modulation; I just mean we could drop the "modulation".

> This sentence is already self-contradicting!?!

You are right. Let me rephrase my thoughts: We are not interested in the attribute "it is an abstraction" but in the attribute "it is an underapproximation of infeasibility" -- or in your words, we are interested in the star of the mercedes, not in it being a car.

> > then calling the whole thing assertion ordering

> What is "the whole thing"

The heuristics. The second part of your question. 

Are you still extremely confused? 

My proposal: 
Call (1) trace reduction, call (2) assertion ordering.

**Comment by Heizmann** [comment]
I dislike both. Maybe you are seeing here some connection that I cannot see. Let us discuss this at the university.
